### ISSUE TITLE
Add call hierarchy feature

### ISSUE THREAD

**Issue by xster** [Dart-Code/Dart-Code#612]
Don't know if it's possible. It's possible to check a function's call hierarchy in IntelliJ. Is it possible to bring it to VSCode?

**Comment by DanTup** [comment]
@bwilkerson I guess this would need support from the analyzer? I can't see anything in the docs that looks like it would provide this.

@xster If we had the data, I'm not sure how we could best display this. We have implemented Type Heirarchy (`F4`?) but render it in a pick list and I think it looks a bit weird (being a tree in a flat list), see https://github.com/Dart-Code/Dart-Code/pull/190#issue-88590653

**Comment by bwilkerson** [comment]
By call hierarchy, do you mean a tree where the top-level nodes are all of the methods containing invocations of a selected method, and the children of each node are the methods containing invocations of the parent node?

If so, then it's all done lazily. Each level in the tree is fetched (by searching for references to an element) when the parent node is expanded. It is done this way both for performance and in order to deal with recursive chains of functions. The result isn't strictly a call hierarchy, it's a reference hierarchy, but the distinction is subtle and not interesting to most users.

**Comment by DanTup** [comment]
Aha, makes sense! I guess we can't wedge it into a picker like we did for types then; it'd have to go in the explorer (tree).

**Comment by xster** [comment]
I wonder if it's possible to augment the `Find all References` peek window and make each descendent node also expandable.

**Comment by DanTup** [comment]
We can't really control this, since Code does all rendering here. However, I think you can do another Find References inside the peek window?

**Comment by DanTup** [comment]
This makes most sense implemented in VS Code - there's an issue at https://github.com/Microsoft/vscode/issues/16110.

**Comment by DanTup** [comment]
VS Code is adding new APIs for first-class support for this, so this might be possible soon :-)

https://code.visualstudio.com/updates/v1_33#_call-hierarchy

**Comment by xster** [comment]
I noticed that in the new release notes too. Woohoo!

**Comment by DanTup** [comment]
@bwilkerson 

I was just having a go at implementing this:

> If so, then it's all done lazily. Each level in the tree is fetched (by searching for references to an element) when the parent node is expanded

The API in VS Code can call us asking for the "incoming calls" and "outgoing calls" for a node (so you can navigate both ways). I think the incoming calls we could get with `search.findElementReferences` (the same we use for "Find References") but I'm not sure if there's something that allows us to get the outgoing calls (eg. for a given function, find the calls it makes out to other functions)?

If there's nothing already there, then we could just build this in LSP on the server (rather than add new APIs that we'll migrate away from) once the LSP spec work for this lands.

**Comment by bwilkerson** [comment]
No, I don't believe that we have a way for clients to ask for the set of elements referenced within a given function/method/initializer body. Adding support specific to LSP sounds reasonable to me; as far as I know none of the clients of the original protocol need this functionality.

**Comment by DanTup** [comment]
Ok, SGTM. It looks like it's tagged for the next LSP version so might not be too long - I'll take look at it there once it arrives. Thanks!

**Comment by bsutton** [comment]
Any update on this one.
It's probably the feature I miss the most coming from java/eclipse.

**Comment by DanTup** [comment]
I have an implementation of this working now. It's not a feature I've really used much myself so I am interested additional testing/feedback once it lands. It's implemented in the Dart SDK (LSP server), so you'd need to use a dev SDK in order for it to show up.

![Screenshot 2022-07-13 at 17 46 08](https://user-images.githubusercontent.com/1078012/178787452-73a6db0b-491e-49b5-baa3-5d393d58c253.png)

One odd thing I've noticed though is that although we supply the ranges of outbound calls to the client, VS Code doesn't seem to use them. When you click a called method in the tree, VS Code jumps to the declaration and never seems to show you the locations of the outbound calls. There's an open issue about this at https://github.com/microsoft/vscode/issues/111454 which may need 👍 's if that's to be addressed :-)

**Comment by bsutton** [comment]
So how do I trigger the call hierarchy?

I'm running dart 2.180.271.0.dev
I'm running dartr code 3.44.0
restarted vscode.
![image](https://user-images.githubusercontent.com/900408/179437389-c46a057e-a680-4ec7-a55b-02739f7fbbd0.png)



I've tried 'find all references' 

[Screencast from 18-07-22 12:33:46.webm](https://user-images.githubusercontent.com/900408/179437306-160a8f3c-1092-4b26-bbf3-e88c9f48ae84.webm)

**Comment by DanTup** [comment]
@bsutton v2.180.271.0.dev looks like a branch made for a beta build (see https://github.com/flutter/flutter/issues/107564) and forked before the change above. When using a newer version that does include the change, Call Hierarchy will appear on the context menu:

<img width="356" alt="Screenshot 2022-07-18 at 11 01 30" src="https://user-images.githubusercontent.com/1078012/179488978-0ea61a1e-ee64-4358-9ad5-066e555fa5a8.png">

**Comment by bsutton** [comment]
@DanTup 
Do you have a version no. that you know should work:
I've tried:
2.18.0-263.0.dev
2.18.0-250.0.dev
2.18.0-271.0.dev

none of which show the menu.

**Comment by bsutton** [comment]
I'm using dswitch to move between dart versions and it only supports the
archive channels so this might have to wait until an archive gets released
with support.

S. Brett Sutton
Noojee Contact Solutions
03 8320 8100

**Comment by DanTup** [comment]
I'm not sure what "archive channels" means in this context, but FWIW the latest Dart beta on https://dart.dev/get-dart/archive is 2.19.0-255.2.beta which should definitely include it. So if you're able to use beta releases (and not only stable releases) then it should be available.
